src/gensxg.cc

```diff
@@ -0,0 +1,307 @@
+#include <string>
+#include <iostream>
+
+#include "libsxg.h"
+
+struct Options {
+ private:
+  static EVP_PKEY* LoadPrivateKey(const std::string& filepath) {
+    FILE* const keyfile = fopen(filepath.c_str(), "r");
+    if (keyfile == nullptr) {
+      return nullptr;
+    }
+    EVP_PKEY* private_key =
+        PEM_read_PrivateKey(keyfile, nullptr, nullptr, nullptr);
+    fclose(keyfile);
+    return private_key;
+  }
+
+  static X509* LoadX509Cert(const std::string& filepath) {
+    FILE* certfile = fopen(filepath.c_str(), "r");
+    if (certfile == nullptr) {
+      return nullptr;
+    }
+    char passwd[] = "";
+    X509* cert = PEM_read_X509(certfile, 0, 0, passwd);
+    fclose(certfile);
+    return cert;
+  }
+
+  static EVP_PKEY* LoadEd25519Pubkey(const std::string& filepath) {
+    FILE* keyfile = fopen(filepath.c_str(), "r");
+    if (!keyfile) {
+      return nullptr;
+    }
+    EVP_PKEY* public_key = PEM_read_PUBKEY(keyfile, nullptr, nullptr, nullptr);
+    fclose(keyfile);
+    return public_key;
+  }
+
+ public:
+  Options(int argc, const char** argv)
+      : help(false),
+        content_type("text/html"),
+        date(0),
+        expire(0),
+        mi_record_size(4096),
+        output("out.sxg") {
+    if (argc <= 2) {
+      help = true;
+      return;
+    }
+    for (int i = 1; i < argc; i += 2) {
+      std::string key(argv[i]);
+      std::string value(argv[i + 1]);
+      if (key == "-help") {
+        help = true;
+        return;
+      } else if (key == "-url") {
+        url = value;
+      } else if (key == "-certUrl") {
+        cert_url = value;
+      } else if (key == "-validityUrl") {
+        validity_url = value;
+      } else if (key == "-certificate") {
+        certificate = value;
+      } else if (key == "-privateKey") {
+        private_key = value;
+      } else if (key == "-content") {
+        content = value;
+      } else if (key == "-contentType") {
+        content_type = value;
+      } else if (key == "-date") {
+        date = std::stoul(value);
+      } else if (key == "-expire") {
+        expire = std::stoul(value);
+      } else if (key == "-miRecordSize") {
+        mi_record_size = std::stoul(value);
+      } else {
+        throw std::runtime_error("Unknown option [" + key + "]");
+      }
+    }
+
+    // Set default values.
+    if (date == 0) {
+      time_t now = time(nullptr);
+      date = static_cast<size_t>(now);
+    }
+    if (expire == 0) {
+      expire = date + 60 * 60 * 24 * 7;  // 7 days.
+    }
+    if (mi_record_size == 0) {
+      mi_record_size = 4096;
+    }
+  }
+
+  std::string Validate() const {
+    std::string ans = "";
+    if (url.empty()) {
+      ans += " -url must not be empty.\n";
+    }
+    if (cert_url.empty()) {
+      ans += " -certUrl must not be empty.\n";
+    }
+    if (validity_url.empty()) {
+      ans += " -validityUrl must not be empty.\n";
+    }
+    if (content.empty()) {
+      ans += " -content file must be specified.\n";
```

Comment:
I think falling back to the standard input is more convenient to users.

Reply:
using `stderr` as error message.